storage: regression test time bounded iterator reuse

tbg opened the pull request:
Release note: None

Without #21721, this fails with

> saw 0 values in regular iterator, but expected 4.

Touches #21721.
Touches #21689.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21723)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/engine/rocksdb_test.go, line 786 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21723#-L3ZtAc7NOF5LOJORuy_:-L3ZtAc8zxKwCpvS1CkL:b-g0fysu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7e56f8df09dcbe2dc7914753484e468c2501b4a/pkg/storage/engine/rocksdb_test.go#L786)):*
> ```Go
> 
> 	// Make a regular iterator. Before #21721, this would accidentally pick up the
> 	// time bounded iterator instead..
> ```

`s/.././g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21723#-:-L3ZtFHzU-ZXesfb_yml:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: regression test time bounded iterator reuse

Release note: None

Without #21721, this fails with

> saw 0 values in regular iterator, but expected 4.

Touches #21721.
Touches #21689.

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/engine/rocksdb_test.go, line 786 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21723#-L3ZtAc7NOF5LOJORuy_:-L3ZtqxmORauWT6ftjUb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa6aff69aa5cd4f5029a2a984e5000810e315d8d/pkg/storage/engine/rocksdb_test.go#L786)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/.././g`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21723)*
<!-- Sent from Reviewable.io -->

knz commented:
nit: a release note can contain multiple paragraphs. As it now stands it looks like the rest of your commit message is part of the release note. To avoid this you can place the release note at the end of the commit message.

bdarnell commented:
This works as a regression test, but I was thinking of a higher-level test that would have some chance of catching the next bug like this before it happens. Something that actually does some splits and verifies the stats afterwards.

tschottdorf commented:
Like the assertions I introduced? You could turn those into testing knobs and just turn them on in some split tests. That said, that's testing a symptom and not the root cause. The consistency queue already verifies the stats (though as we know the stats can always be wrong on real clusters right now, so it doesn't yet complain).

bdarnell commented:
Yeah, I'd like to have those assertions turned on in at least some tests. I don't look at it as a symptom vs root cause thing so much as an integration test vs unit testing. It's only by testing the symptoms that you can detect problems whose root causes you don't know to look for yet.
